tests/unit/test_agw_configurator.py

```diff
@@ -0,0 +1,169 @@
+        )
+        self.agw_configurator = AGWConfigurator(self.TEST_DOMAIN, self.TEST_ROOT_CA_PEM_PATH)
+
+    @patch("magma_access_gateway_configurator.agw_configurator.os.makedirs")
+    @patch(
+        "magma_access_gateway_configurator.agw_configurator.shutil.copy",
+        Mock(),
+    )
+    def test_given_non_existent_root_ca_pem_destination_dir_when_copy_root_ca_pem_then_root_ca_pem_destination_dir_is_created(  # noqa: E501
+        self, mocked_makedirs
+    ):
+        with self.mocked_root_ca_pem_dest_dir:
+            self.agw_configurator.copy_root_ca_pem()
+
+        mocked_makedirs.assert_called_once_with(self.TEST_ROOT_CA_PEM_DESTINATION_DIR)
+
+    @patch("magma_access_gateway_configurator.agw_configurator.os.makedirs")
+    @patch(
+        "magma_access_gateway_configurator.agw_configurator.os.path.exists",
+        Mock(return_value=True),
+    )
+    @patch(
+        "magma_access_gateway_configurator.agw_configurator.shutil.copy",
+        Mock(),
+    )
+    def test_given_existent_root_ca_pem_destination_dir_when_copy_root_ca_pem_then_root_ca_pem_destination_dir_is_not_created(  # noqa: E501
+        self, mocked_makedirs
+    ):
+        with self.mocked_root_ca_pem_dest_dir:
+            self.agw_configurator.copy_root_ca_pem()
+
+        mocked_makedirs.assert_not_called()
+
+    @patch("magma_access_gateway_configurator.agw_configurator.shutil.copy")
+    @patch(
+        "magma_access_gateway_configurator.agw_configurator.os.path.exists",
+        Mock(return_value=True),
+    )
+    def test_given_existent_root_ca_pem_destination_dir_when_copy_root_ca_pem_then_root_ca_pem_is_copied_to_the_destination_dir(  # noqa: E501
+        self, mocked_copy
+    ):
+        with self.mocked_root_ca_pem_dest_dir, self.mocked_root_ca_pem_file_name:
+            self.agw_configurator.copy_root_ca_pem()
+
+        mocked_copy.assert_called_once_with(
+            self.TEST_ROOT_CA_PEM_PATH,
+            os.path.join(self.TEST_ROOT_CA_PEM_DESTINATION_DIR, self.TEST_ROOT_CA_PEM_FILE_NAME),
+        )
+
+    @patch("magma_access_gateway_configurator.agw_configurator.os.makedirs")
+    @patch("magma_access_gateway_configurator.agw_configurator.os.path.exists")
+    def test_given_non_existent_control_proxy_conf_dir_when_configure_control_proxy_then_control_proxy_conf_dir_is_created(  # noqa: E501
+        self, mocked_path_exists, mocked_makedirs
+    ):
+        mocked_path_exists.side_effect = [False, True]
+
+        with self.mocked_magma_control_proxy_config_dir:
+            self.agw_configurator.configure_control_proxy()
+
+        mocked_makedirs.assert_called_once_with(self.TEST_MAGMA_CONTROL_PROXY_CONFIG_DIR)
+
+    @patch("magma_access_gateway_configurator.agw_configurator.os.makedirs")
+    @patch("magma_access_gateway_configurator.agw_configurator.os.path.exists")
+    def test_given_existent_control_proxy_conf_dir_when_configure_control_proxy_then_control_proxy_conf_dir_is_not_created(  # noqa: E501
```

Comment:
Why is the reason for this parameter order, why it shouldn't be `self, mocked_makedirs, mocked_path_exists`, as the first patch is `makedirs`. What am I missing here?

Reply:
Nevermind

Reply:
My bad